### PR TITLE
Feature/net p2p pingback

### LINKED ISSUE #1581: DeserializeError on calls to /v2/info
**Describe the bug**
Calling `/v2/info` from a browser results in a DeserializeError in the logs. Seems like it's due to the favicon request that browsers send.

**Steps To Reproduce**
 - Run a node
 - Open `http://127.0.0.1:20443/v2/info` in a browser

**Expected behavior**
No errors / warnings in logs.

**Environment (please complete the following information):**
 - OS: [e.g. Ubuntu / Debian]: MacOS
 - git-rev ae9ed9af1

**Additional context**

Relevant logs:
```
INFO [1589482740.041] [src/net/http.rs:2377] Failed to parse HTTP request: DeserializeError("Http request could not be parsed")
DEBUG [1589482740.041] [src/net/connection.rs:553] Invalid message payload: DeserializeError("Http request could not be parsed").  Preamble was Request(HttpRequestPreamble { version: Http11, verb: "GET", path: "/favicon.ico", host: PeerHost::IP(00000000000000000000ffff7f000001,20443), content_type: None, content_length: None, keep_alive: true, headers: {"user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36", "referer": "http://127.0.0.1:20443/v2/info", "accept": "image/webp,image/apng,image/*,*/*;q=0.8", "sec-fetch-dest": "empty", "accept-encoding": "gzip, deflate, br", "sec-fetch-site": "same-origin", "accept-language": "en-US,en;q=0.9", "dnt": "1", "sec-fetch-mode": "no-cors"} })
DEBUG [1589482740.042] [src/net/rpc.rs:991] http:id=4,request=false: failed to recv: InvalidMessage
DEBUG [1589482740.042] [src/net/server.rs:352] Got a bad HTTP message on socket TcpStream { addr: V4(127.0.0.1:20443), peer: V4(127.0.0.1:55877), fd: 5 }
DEBUG [1589482740.042] [src/net/server.rs:528] Close HTTP connection on event 4
```

### PR DESCRIPTION
**Pull request by jcnelson**
This PR adds two new features to the peer-to-peer networking code: **pingbacks** and **inbound neighbor walks**.

Pingbacks are when a node tries to handshake with another node that has recently handshaked with it.  This causes each publicly-routable node to add a directed edge to all other publicly-routable peers that talk to it (i.e. who are the origins of the node's inbound edges), thereby ensuring that there exists a bi-directional edge in the peer graph between any pair (A, B) of publicly routable nodes such that A connects to B _or_ B connects to A.

Inbound neighbor walks are when a node asks a peer with an inbound edge for its neighbors, and then proceeds to try to handshake with them.  This causes nodes that are publicly routable to ask nodes that may _not_ be publicly routable for their sets of publicly-routable nodes, thereby ensuring that there will exist a bi-directional edge in the peer graph between any pair (A, B) of publicly routable nodes if there exists distinct node C such that edges `C --> A` and `C --> B` exist (the existence of these edges will cause `A --> B` and `B --> A` to come into existence once A or B does an inbound neighbor walk to C).

If a node's maximum number of inbound and outbound edges are infinite, then these changes ensure that if the initial peer graph is connected, then the peer graph will evolve to the point where there exists a bidirectional edge between every publicly-routable node, and a directional edge from each non-publicly-routable node (i.e. those behind NATs) to every publicly routable node.  The graph would look like two concentric rings -- a "core" ring of publicly-routable nodes that are all-to-all connected, and a "peripheral" ring of non-publicly-routable nodes that are connected to all publicly-routable nodes.  This is the ideal network topology we want.

In practice, the number of inbound and outbound edges are capped, so the resulting peer graph will consist of the same vertices (nodes), but with a random subset of the edges.

EDIT: this PR updates `GET /v2/neighbors` to return not only a random frontier sample, but also a list of inbound and outbound peers.  This should make it easier external tooling to monitor the status of the peer network as we're rolling it out.

EDIT: this PR also fixes #1581 by adding a `HttpRequestType::Unmatched()` variant to capture unmatched paths (which the RPC server handles with a `HttpResponsType::NotFound()`).

## How to Test

To test the inbound neighbor walk feature, you'd spin up two nodes: a "remote" node that your laptop can contact (doesn't have to be publicly-routable), and a "local" node that has both this "remote" node and the Neon master node as its initial peers.  You should see the "remote" node ask the "local" node for neighbors, and in so doing, discover the Neon master node.  You may or may not see a pingback occur if you look at the logs.

### CODE REVIEW

File: src/net/rpc.rs [hunk 1]
```diff
@@ -163,6 +166,38 @@ impl PeerInfoData {
...
 }
 
+impl RPCNeighborsInfo {
+    /// Load neighbor address information from the peer network
+    pub fn from_p2p(network_id: u32, peers: &PeerMap, chain_view: &BurnchainView, peerdb: &PeerDB) -> Result<RPCNeighborsInfo, net_error> {
+        let neighbor_sample = PeerDB::get_random_neighbors(peerdb.conn(), network_id, MAX_NEIGHBORS_DATA_LEN, chain_view.burn_block_height, false)
+            .map_err(net_error::DBError)?;
+
+        let sample : Vec<RPCNeighbor> = neighbor_sample
+            .iter()
```
Comment (kantai) [inline src/net/rpc.rs:176]: You can avoid a clone here by using `into_iter`

Comment (jcnelson) [reply src/net/rpc.rs]: Fixed in 5dda11c

File: src/net/neighbors.rs [hunk 2]
```diff
@@ -57,16 +58,13 @@ use burnchains::BurnchainView;
...
 use rand::prelude::*;
 use rand::thread_rng;
 
-use rusqlite::Transaction;
-
 #[cfg(test)] pub const NEIGHBOR_MINIMUM_CONTACT_INTERVAL : u64 = 0;
 #[cfg(not(test))] pub const NEIGHBOR_MINIMUM_CONTACT_INTERVAL : u64 = 600;      // don't reach out to a frontier neighbor more than once every 10 minutes
 
-pub const NEIGHBOR_REQUEST_TIMEOUT : u64 = 10;
+pub const NEIGHBOR_REQUEST_TIMEOUT : u64 = 30;      // number of seconds an outstanding request for neighbors can take
```
Comment (kantai) [inline src/net/neighbors.rs:67]: Should this be configurable?

File: src/net/neighbors.rs [hunk 3]
```diff
@@ -147,7 +157,9 @@ impl Neighbor {
     /// upper bound.
     pub fn degree(&self) -> u64 {
         let mut rng = thread_rng();
-        let res = rng.gen_range(self.in_degree, self.out_degree+1) as u64;
+        let min = if self.in_degree < self.out_degree { self.in_degree } else { self.out_degree };
+        let max = if self.in_degree >= self.out_degree { self.in_degree } else { self.out_degree };
```
Comment (kantai) [inline src/net/neighbors.rs:161]: `cmp::min` and `cmp::max` are handy for this kind of pattern.

File: src/net/neighbors.rs (hunk 2 above)
Comment (jcnelson) [reply src/net/neighbors.rs:67]: sure; I can add that.

**Pull request merged**